### ISSUE TITLE
how to modify request in middleware?

### ISSUE THREAD

**Issue by whichdew** [tiangolo/fastapi#5165]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.middleware("http")
async def add_request_user_to_body(req: Request, call_next):
    '''
    description: 
    return {*}
    '''
    username=req.headers['username']
    if req.method.lower()!='get':
        req._body['username']=username
    response = await call_next(req)
    return response
```

### Description

1. i use jwt for auth, client will carry its jwt in headers everytime.
2. i can get username from jwt and use it as data owner.
3. for convenience, i want to add username to body which is from jwt. then i can use body.dict() to init my db model directly.
4. so, i want to achieve it in middleware instead of on each route.

then how to modify request to add custom data？

### Operating System

Linux

### FastAPI Version

0.73.0

### Python Version

3.9

**Comment by JarroVGIT** [comment]
> is it correct to modify request.state like `request.state.username=username` ?

You can change `request.state` as you please and that is totally fine, for example like `request.state["username"] = 'username1'`.

To get this information in your path operation function (it is not called a route handler), you can do this:

```python
from fastapi import FastAPI, Request

@router.post('/test')
async def handler(request: Request, data:MyModel):
    usr = request.state['username']
    return usr
```

**Comment by JarroVGIT** [comment]
I had to edit te above as it was incomplete, my apologies

**Comment by JarroVGIT** [comment]
Sorry one last comment, as you explicitly asked for a way to do this in middleware. Below is a fully working example on **both** approaches (mw and dependency):

```python
from fastapi import FastAPI, HTTPException, Request, Depends

async def username_dependency(r: Request):
    #put your JWT logic here, I will just assume you retrieved your username
    r.state.is_authenticated_from_dependency = True
    r.state.username_from_dependency = "user_from_dependency"
    return

app = FastAPI(dependencies=[Depends(username_dependency)])

@app.middleware("http")
async def add_request_user_to_body(r: Request, call_next):
    #put your JWT logic here, I will just assume you retrieved your username
    r.state.is_authenticated_from_mw= True
    r.state.username_from_mw = "user_from_middleware"
    response = await call_next(r)
    return response

@app.get("/from_dependency")
async def root(r: Request):
    if r.state.is_authenticated_from_dependency:
        return r.state.username_from_dependency
    return HTTPException(401, "You are not authenticated!")

@app.get("/from_middleware")
async def root(r: Request):
    if r.state.is_authenticated_from_mw:
        return r.state.username_from_mw
    return HTTPException(401, "You are not authenticated!")

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

**Comment by whichdew** [comment]
> Sorry one last comment, as you explicitly asked for a way to do this in middleware. Below is a fully working example on **both** approaches (mw and dependency):
> 
> ```python
> from fastapi import FastAPI, HTTPException, Request, Depends
> 
> async def username_dependency(r: Request):
>     #put your JWT logic here, I will just assume you retrieved your username
>     r.state.is_authenticated_from_dependency = True
>     r.state.username_from_dependency = "user_from_dependency"
>     return
> 
> app = FastAPI(dependencies=[Depends(username_dependency)])
> 
> @app.middleware("http")
> async def add_request_user_to_body(r: Request, call_next):
>     #put your JWT logic here, I will just assume you retrieved your username
>     r.state.is_authenticated_from_mw= True
>     r.state.username_from_mw = "user_from_middleware"
>     response = await call_next(r)
>     return response
> 
> @app.get("/from_dependency")
> async def root(r: Request):
>     if r.state.is_authenticated_from_dependency:
>         return r.state.username_from_dependency
>     return HTTPException(401, "You are not authenticated!")
> 
> @app.get("/from_middleware")
> async def root(r: Request):
>     if r.state.is_authenticated_from_mw:
>         return r.state.username_from_mw
>     return HTTPException(401, "You are not authenticated!")
> 
> if __name__ == "__main__":
>     import uvicorn
>     uvicorn.run(app, host="0.0.0.0", port=8000)
> ```

thanks a lot. however, as you can see, if i want to use username in restful-api like below, i have to repeat some code in every path operation function. is there any way to modify request.body directly?

```python
class MyModel(BaseModel):
    id:int
    owner:str # username
    value:Any
    
@router.post('/api/{table_name}/')
async def restful_api(table_name,data:MyModel,req:Request):
    
    # the following 3 lines will repeat in put、patch、delete
    # if there is a way so that i can set data.username in middleware/dependcy, it would be better
    if not is_authencated:
        raise Exception()
    data.username=req.state.username

    session=get_table_session(table_name)
    session.insert(data)
   return data.id

```

**Comment by JarroVGIT** [comment]
You can combine stuff together to get to such a solution like the below. It does the authentication stuff in middleware, and access it in the `__init__` of a dependable class. 

```python
from typing import Any
from fastapi import Body, FastAPI, HTTPException, Request, Depends

class RequestModel:
    def __init__(self, r: Request, id: int = Body(), value: Any = Body()) -> None:
        self.id = id
        self.value = value
        if r.state.is_authenticated:
            self.id = r.state.username
        else:
            raise HTTPException(401, "Not Authorized")

app = FastAPI()

@app.middleware("http")
async def add_request_user_to_body(r: Request, call_next):
    #put your JWT logic here, I will just assume you retrieved your username
    r.state.is_authenticated = True
    r.state.username = "user_from_middleware"
    response = await call_next(r)
    return response

@app.post("/test")
async def root(r: Request, data: RequestModel = Depends() ):
    return {"Hello": data.id}


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

Calling the endpoint:
```bash
curl -X 'POST' \
  'http://0.0.0.0:8000/test' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "id": 3,
  "value": "someValue"
}'
```

Results in:
```json
{
  "Hello": "user_from_middleware",
  "YourID": 3
}
```

**Comment by whichdew** [comment]
> You can combine stuff together to get to such a solution like the below. It does the authentication stuff in middleware, and access it in the `__init__` of a dependable class.
> 
> ```python
> from typing import Any
> from fastapi import Body, FastAPI, HTTPException, Request, Depends
> 
> class RequestModel:
>     def __init__(self, r: Request, id: int = Body(), value: Any = Body()) -> None:
>         self.id = id
>         self.value = value
>         if r.state.is_authenticated:
>             self.username = r.state.username
>         else:
>             raise HTTPException(401, "Not Authorized")
> 
> app = FastAPI()
> 
> @app.middleware("http")
> async def add_request_user_to_body(r: Request, call_next):
>     #put your JWT logic here, I will just assume you retrieved your username
>     r.state.is_authenticated = True
>     r.state.username = "user_from_middleware"
>     response = await call_next(r)
>     return response
> 
> @app.post("/test")
> async def root(data: RequestModel = Depends() ):
>     return {"Hello": data.username, "YourID": data.id}
> 
> 
> if __name__ == "__main__":
>     import uvicorn
>     uvicorn.run(app, host="0.0.0.0", port=8000)
> ```
> 
> Calling the endpoint:
> 
> ```shell
> curl -X 'POST' \
>   'http://0.0.0.0:8000/test' \
>   -H 'accept: application/json' \
>   -H 'Content-Type: application/json' \
>   -d '{
>   "id": 3,
>   "value": "someValue"
> }'
> ```
> 
> Results in:
> 
> ```json
> {
>   "Hello": "user_from_middleware",
>   "YourID": 3
> }
> ```

 A great work-around, thank you!

 It seems that it's complex to modify `request.body` directly. I will keep exploring.

**Comment by JarroVGIT** [comment]
Why would you want to change the request body itself, while you can specifically put logic into the State object of a request? I'd say it is a worse practice to manually change the user input, rather than separating the input and derived information from that input.

**Comment by whichdew** [comment]
> Why would you want to change the request body itself, while you can specifically put logic into the State object of a request? I'd say it is a worse practice to manually change the user input, rather than separating the input and derived information from that input.

From the point of view of design, what you said is indeed right. One thing I'm adamant about is that I want to simplify the code, which means making a lot of use of reflection and automation for convention based programming. In my current project, the data format of interaction between client and server is determined, and this way of modifying request only applies to some routes, such as RESTFUL API, so modifying request is the easiest for me.

**Comment by iudeen** [comment]
May I suggest a different approach? 

I know you asked for altering the request body. Yes, it can be done, but can be bit complex to start with.

Before getting into that, how about if I suggest you use dependency injection rather than a Middleware to pass username to your routes.

**Comment by whichdew** [comment]
> 

yath, I accept it. dependency indeed is  better.

**Comment by nikoladsp** [comment]
Hi all,
in order not to open another similar question, I will be free to ask here (hope not a problem).

I would like to use refresh JWT token in my app. So all endpoints will use access token (short lifespan) in Bearer field. Only exceptions are login endpoint (no token needed at all) and endpoint where I want to generate access token itself - in which case refresh token is sent in Bearer field. 

There are different signing keys in these two cases: when coding/decoding refresh token, I am using application key (env variable, something does not change during runtime) and for coding/decoding access token I am using refresh token which is generated _**per user**_ (each user has its own)

My questions are as follows:
1. how can I (if possible) set dependency for majority of endpoints in FastAPI constructor, so I don't have to drag the same code everywhere and put exception/override on couple of specific ones?
2. my dependency should obtain SqlAlchemy session and also a user object pre-loaded so they can be passed to different services called from endpoints. What would be the recommended way? If I want to use Swagger, shall I inherit my dependency class from `HTTPBearer` and return `credentials.credentials`? Does this class must have

    ```python
    async def __call__(self, request: Request):
    ```
    This part is bit unclear to me. Can/shall I use the above approach or if there is a better/cleaner in my case?

Thank you kindly.

Best regards

**Comment by iudeen** [comment]
@nikoladsp I think its needed since you'd be inheriting a base class. 

```python
    def __call__(self, request: Request, response: Response) -> str:
        cookies = request.cookies
        login_token = cookies.get("login-token")
        if not login_token:
            login_token = request.headers.get("login-token")
        if not login_token:
            raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED)

       # --------- some magic sauce --------------
       return user_id

```

**Comment by iudeen** [comment]
yes! see this [documentation](https://fastapi.tiangolo.com/tutorial/sql-databases/#create-a-dependency)

You need to create another dependency for managing your session per request.

**Comment by iudeen** [comment]
> @iudeen
> 
> Thank you.
> 
> Is there a good practice to create a DB session and "attach" it to Request (or some other object), in such way I close it when request is finished, so I don't leak connections?

yes! see this [documentation](https://fastapi.tiangolo.com/tutorial/sql-databases/#create-a-dependency)

You need to create another dependency for managing your session per request.

**Comment by nikoladsp** [comment]
Great!
